src/v/security/audit/schemas/application_activity.h

```diff
@@ -120,4 +121,56 @@ class api_activity final : public ocsf_base_event {
         w.EndObject();
     }
 };
+
+class application_lifecycle final : public ocsf_base_event {
+public:
+    enum class activity_id : uint8_t {
+        unknown = 0,
+        install = 1,
+        remove = 2,
+        start = 3,
+        stop = 4,
+        other = 99
+    };
+
+    application_lifecycle(
+      activity_id activity_id,
+      product&& app,
+      enum severity_id severity_id,
+      timestamp_t time)
+      : ocsf_base_event(
+        category_uid::application_activity,
+        class_uid::application_lifecycle,
+        severity_id,
+        time,
+        activity_id)
+      , activity_id(activity_id)
+      , app(std::move(app)) {
+        _key = hash();
+    }
+
+    auto equality_fields() const {
```

Comment:
i wonder if there's a way to make this less error prone. I'm imagining these sub classes just return a tuple of the types they are interested in hashing. For example here:
```
auto equality_fields() const { return std::tie(activity_id, app); }
```
Then one base class method combines them all. That way an implementer of a child class cannot forget about tying the `base_equality_fields` values into their `equality_fields` result.

Reply:
The issue is that the base class can't call any methods in the child class during construction (which would be where I think we would like for this to occur).

Reply:
yeah i wouldn't store the hash in memory either, solving this issue

Reply:
If you're doing it that way, there's no need to use polymorphism, but I guess at some point these will end up in a container and you don't want to use a variant?

You could make the call to hash lazy, use memoization, and implement template method (with non-virtual public interface):
```c++
class base {
public:
    void hash() const {
        if (!hash.hash_value()) {
            _hash.emplace(boost::hash_combine(std::hash(equality_fields()), do_hash());
        }
        return _hash.value();
    }
private:
    // Only hash your derived fields
    virtual size_t do_hash() = 0;
    std::optional<size_t> _hash;
};
```

Reply:
You could make the call to hash lazy, use memoization, and implement template method (with non-virtual public interface):
```c++
class base {
public:
    void hash() const {
        if (!hash.hash_value()) {
            _hash.emplace(boost::hash_combine(std::hash(equality_fields()), do_hash());
        }
        return _hash.value();
    }
private:
    // Only hash your derived fields
    virtual size_t do_hash() = 0;
    std::optional<size_t> _hash;
};
```

Reply:
@BenPope yeah was thinking something like that